kafka.go

```diff
@@ -1,161 +1,213 @@
-	for err := range producer.Errors() {
-		logrus.Error("unable to push event to Kafka stream: ", err)
+// Publish send event to message broker
+func (client *KafkaClient) Publish(publishBuilder *PublishBuilder) {
+	message, err := constructEvent(publishBuilder)
+	if err != nil {
+		logrus.Errorf("unable to construct event : %s , error : %v", publishBuilder.eventName, err)
+		return
 	}
-}
 
-type eventEntry struct {
-	Event
+	for _, topic := range publishBuilder.topic {
+		config := client.publishConfig
+		config.Topic = topic
+		writer := kafka.NewWriter(config)
 
-	encoded []byte
-	err     error
+		err = writer.WriteMessages(publishBuilder.ctx, message)
+		if err != nil {
+			logrus.Errorf("unable to publish event to kafka. topic : %s , error %v : ", topic, err)
+		}
+	}
 }
 
-func (evtEntry *eventEntry) ensureEncoded() {
-	if evtEntry.encoded == nil && evtEntry.err == nil {
-		evtEntry.encoded, evtEntry.err = json.Marshal(evtEntry)
+// constructEvent construct event message
+func constructEvent(publishBuilder *PublishBuilder) (kafka.Message, error) {
+	id := generateID()
+	event := &Event{
+		ID:        id,
+		EventName: publishBuilder.eventName,
+		Namespace: publishBuilder.namespace,
+		ClientID:  publishBuilder.clientID,
+		UserID:    publishBuilder.userID,
+		Timestamp: time.Now(),
+		Version:   version,
+		Payload:   publishBuilder.payload,
 	}
-}
 
-// Length returns the length of the event entry
-func (evtEntry *eventEntry) Length() int {
-	evtEntry.ensureEncoded()
-	return len(evtEntry.encoded)
-}
+	eventBytes, err := marshal(event)
+	if err != nil {
+		return kafka.Message{}, err
+	}
+
+	return kafka.Message{
+		Key:   []byte(id),
+		Value: eventBytes,
+	}, nil
 
-// Encode serialize the struct
-func (evtEntry *eventEntry) Encode() ([]byte, error) {
-	evtEntry.ensureEncoded()
-	return evtEntry.encoded, evtEntry.err
 }
 
-// PublishEvent push an event to a Kafka topic synchronously
-func (client *KafkaClient) PublishEvent(event *Event) error {
-	event.Realm = client.realm
-	event.Time = time.Now().UTC()
+// Register register callback and consume a topic
+func (client *KafkaClient) Register(subscribeBuilder *SubscribeBuilder) {
+	topic := constructTopic(client.realm, subscribeBuilder.topic)
+	if client.registerCallback(topic, subscribeBuilder.eventName, subscribeBuilder.callback) {
+		logrus.Warnf("topic and event already registered. event : %s", subscribeBuilder.eventName)
+		return
+	}
 
-	_, _, err := client.syncProducer.SendMessage(&sarama.ProducerMessage{
-		Timestamp: time.Now().UTC(),
-		Value: &eventEntry{
-			Event: *event,
-		},
-		Topic: event.Topic,
-	})
+	config := client.subscribeConfig
+	config.Topic = subscribeBuilder.topic
+	reader := kafka.NewReader(config)
+	if err := reader.SetOffset(kafka.LastOffset); err != nil {
+		logrus.Error("unable to set offset : ", err)
+		return
+	}
 
-	return err
+	// listening a topic
+	// each topic would have his own goroutine
+	go func() {
+		defer func() {
+			_ = reader.Close()
```

Comment:
Is there any reason we put the deferred reader.Close inside the goroutine?

Reply:
the defer should be perform when the consumer got error and because consumer working on other goroutine, so we register the it inside of consumer goroutine
but, actually register / consume topic is non-blocking function so we can put all step inside of goroutine

Reply:
if register / consume is non-blocking, then why do you still need to call everything in goroutine?

Reply:
because each goroutine it used for specific topic, so we need to close the reader if got something error when consuming the topic.

Reply:
oh I see, I missed the loop below